contracts/modules/elrond-wasm-module-governance/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+#![no_std]
+
+elrond_wasm::imports!();
+
+pub mod governance_configurable;
+
+pub mod governance_proposal;
+use governance_proposal::*;
+
+#[elrond_wasm_derive::module]
+pub trait GovernanceModule:
+	governance_configurable::GovernanceConfigurablePropertiesModule
+{
+	// TODO: Withdraw functionality after voting ends
+
+	// endpoints
+
+	// Used to deposit tokens for "payable" actions
+	// There is no "withdraw" functionality
+	// Funds can only be retrived through an action
+	#[payable("*")]
+	#[endpoint(depositTokensForAction)]
+	fn deposit_tokens_for_action(&self) {}
+
+	// Used to withdraw the tokens after the action was executed or cancelled
+	#[endpoint(withdrawGovernanceTokens)]
+	fn withdraw_governance_tokens(&self, proposal_id: usize) -> SCResult<()> {
+		self.require_valid_proposal_id(proposal_id)?;
+		require!(
+			self.get_proposal_status(proposal_id) == GovernanceProposalStatus::None,
+			"Proposal has to be executed or canceled first"
+		);
+
+		let caller = self.blockchain().get_caller();
+		let governance_token_id = self.governance_token_id().get();
+		let nr_votes_tokens = self.votes(proposal_id).get(&caller).unwrap_or_default();
+		let nr_downvotes_tokens = self.downvotes(proposal_id).get(&caller).unwrap_or_default();
+
+		if nr_votes_tokens > 0 {
+			self.send()
+				.direct(&caller, &governance_token_id, &nr_votes_tokens, &[]);
+
+			self.votes(proposal_id).remove(&caller);
```

Comment:
It's a best practice to clear the storage and then send the tokens. Right now there's no risk of re-entrancy since there's no function call, but still if someone decides to add, he should not worry about this risk

Reply:
Fixed.